tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11236,1074 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  LongVectorOpTestBase<uint32_t>(LongVectorOpType_Initialize);
+}
+
+TEST_F(ExecutionTest, LongVector_Clamp_uint64) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  LongVectorOpTestBase<uint64_t>(LongVectorOpType_Clamp);
+}
+
+TEST_F(ExecutionTest, LongVector_Initialize_uint64) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  LongVectorOpTestBase<uint64_t>(LongVectorOpType_Initialize);
+}
+
+template <typename T>
+void ExecutionTest::LongVectorOpTestBase(LongVectorOpType opType) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  LongVectorOpTestConfig<T> TestConfig(opType);
+
+  LongVectorOpTestBase<T, 3>(TestConfig);
+  LongVectorOpTestBase<T, 4>(TestConfig);
+  LongVectorOpTestBase<T, 5>(TestConfig);
+  LongVectorOpTestBase<T, 16>(TestConfig);
+  LongVectorOpTestBase<T, 17>(TestConfig);
+  LongVectorOpTestBase<T, 35>(TestConfig);
+  // TODOLONGVEC: 100, 1024 breaks the size limit for structured buffers
+  // LongVectorOpTestBase<T, 100>(TestConfig);
+  // LongVectorOpTestBase<T, 1024>(TestConfig);
+}
+
+template <typename T, std::size_t N>
+void ExecutionTest::LongVectorOpTestBase(
+    LongVectorOpTestConfig<T> &TestConfig) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  LogCommentFmt(L"Running LongVectorOpTestBase<%S, %zu>", typeid(T).name(), N);
+
+  CComPtr<ID3D12Device> D3DDevice;
+  if (!CreateDevice(&D3DDevice, D3D_SHADER_MODEL_6_9)
+      && !m_ExperimentalModeEnabled) {
+    if(m_HLKModeEnabled)
+      LogErrorFmt(L"Device does not support SM 6.9. Can't run these tests.");
+      return;
+
+    WEX::Logging::Log::Comment(
+        "Device does not support SM 6.9. Can't run these tests.");
+    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+    return;
+  }
+
+  DeterministicNumberGenerator<T> NumberGenerator(1337);
+  std::vector<T> VecInput1 = {};
+  std::vector<T> VecInput2 = {};
+  const T ScalarInput = NumberGenerator.generate();
+  const bool IsVectorBinaryOp = TestConfig.IsBinaryOp && !TestConfig.IsScalarOp;
+
+  // Fill the vector inputs with values.
+  for (size_t i = 0; i < N; i++) {
+    // Always generate input. Duh.
+    VecInput1.push_back(NumberGenerator.generate());
+
+    if (IsVectorBinaryOp) {
+      VecInput2.push_back(NumberGenerator.generate());
+    }
+  }
+
+  // We pass these values into the shader and they're requried to compile. So
+  // they need to set to something.
+  T ClampArgC = 0;
+  T ClampArgT = 0;
+  if (TestConfig.IntrinsicString == "testClamp") {
+    ClampArgC = NumberGenerator.generate();
+    ClampArgT = NumberGenerator.generate();
+    while (ClampArgC >= ClampArgT) {
```

Comment:
Could this infinitely loop if T is a bool? maybe the 1337 seed works out so this never happens. Might want a case where if its bool, then just directly force the values. Unless you're excluding clamp for bools.

Separately, nit, the comment should be smaller or equal, or you should adjust the condition so its ClampArgC > ClampArgT

Reply:
Good catch. In theory there is probably a seed when T is bool that could lead to an infinite loop if the condition was only >. I think it should be fine if we stick with the >= though?

Also, looking at this. I broke this logic because I changed the string to "TestClamp" instead of "testClamp". I'll swap to comparing the OpType.

Reply:
Assuming the only valid values for bools are 0 and 1, 
(because of
else if constexpr (std::is_same_v<T, HLSLBool_t>)
      // Anything non-zero, including negative values, is true as a bool. So 0
      // and 1 are all we need.
      Uint16Dist = std::uniform_int_distribution<uint16_t>(0, 1);
)
and assuming that ClampArgC is set to 1 initially,
it will never be the case that you could generate a value for ClampArgT that is strictly greater than ClampArgC.
You'll be generating values forever.
Whether the condition is > or >=, both will infinitely loop forever if ClampArgC gets 1.
And IMO clamping doesn't make much sense for bools, so you might consider just forcing these values in the bool case.